File: agents/base_agent.py
"""
Base Agent for ERC-8004 Registry Interactions

This module provides the foundational class for agents that interact with
the ERC-8004 registry contracts in the ZK rebalancing system.
"""

import json
import os
from typing import Dict, Optional, Any
from web3 import Web3
from dotenv import load_dotenv

load_dotenv()


class ERC8004BaseAgent:
    """Base class for agents interacting with ERC-8004 registries"""

    def __init__(self, agent_domain: str, private_key: str):
        """
        Initialize the base agent

        Args:
            agent_domain: The domain where this agent's AgentCard is hosted
            private_key: Private key for signing transactions
        """
        self.agent_domain = agent_domain
        self.private_key = private_key

        # Initialize Web3 connection
        rpc_url = os.getenv("RPC_URL", "http://127.0.0.1:8545")
        self.w3 = Web3(Web3.HTTPProvider(rpc_url))

        if not self.w3.is_connected():
            raise ConnectionError(f"Failed to connect to {rpc_url}")

        # Load account from private key
        self.account = self.w3.eth.account.from_key(private_key)
        self.address = self.account.address

        # Load contract addresses from deployment
        self._load_contract_addresses()

        # Initialize contract instances
        self._init_contracts()

        # Agent registry info
        self.agent_id: Optional[int] = None
        self._check_registration()

    def _load_contract_addresses(self):
        """Load contract addresses from deployed_contracts.json"""
        deployment_path = os.path.join(
            os.path.dirname(__file__), "..", "deployed_contracts.json"
        )
        
        if not os.path.exists(deployment_path):
            # Try alternate path
            deployment_path = "deployed_contracts.json"
        
        try:
            with open(deployment_path, "r") as f:
                deployment = json.load(f)
                contracts = deployment["contracts"]

                self.identity_registry_address = contracts["IdentityRegistry"]
                self.reputation_registry_address = contracts["ReputationRegistry"]
                self.validation_registry_address = contracts["ValidationRegistry"]
        except FileNotFoundError:
            raise FileNotFoundError(
                "deployed_contracts.json not found. Please deploy contracts first."
            )

    def _load_contract_abi(self, contract_name: str) -> list:
        """Load contract ABI from compiled artifacts"""
        abi_path = f"contracts/out/{contract_name}.sol/{contract_name}.json"

        with open(abi_path, "r") as f:
            artifact = json.load(f)
            return artifact["abi"]

    def _init_contracts(self):
        """Initialize contract instances"""
        # Load ABIs
        identity_abi = self._load_contract_abi("IdentityRegistry")
        reputation_abi = self._load_contract_abi("ReputationRegistry")
        validation_abi = self._load_contract_abi("ValidationRegistry")

        # Create contract instances
        self.identity_registry = self.w3.eth.contract(
            address=self.identity_registry_address, abi=identity_abi
        )

        self.reputation_registry = self.w3.eth.contract(
            address=self.reputation_registry_address, abi=reputation_abi
        )

        self.validation_registry = self.w3.eth.contract(
            address=self.validation_registry_address, abi=validation_abi
        )

    def _check_registration(self):
        """Check if this agent is already registered"""
        try:
            result = self.identity_registry.functions.resolveByAddress(
                self.address
            ).call()
            if result[0] > 0:  # AgentID > 0 means registered
                stored_domain = result[1]
                if stored_domain == self.agent_domain:
                    self.agent_id = result[0]
                    print(f"✅ Agent already registered with ID: {self.agent_id}")
                else:
                    print(f"⚠️  Address registered with different domain: {stored_domain}")
                    print("ℹ️  Agent not yet registered with current domain")
            else:
                print("ℹ️  Agent not yet registered")
        except Exception as e:
            error_str = str(e)
            if "0xe93ba223" in error_str or "AgentNotFound" in error_str:
                print("ℹ️  Agent not yet registered")
            else:
                print(f"ℹ️  Error checking registration: {e}")

    def register_agent(self) -> int:
        """
        Register this agent with the IdentityRegistry

        Returns:
            Agent ID assigned by the registry
        """
        if self.agent_id:
            print(f"Agent already registered with ID: {self.agent_id}")
            return self.agent_id

        print(f"📝 Registering agent with domain: {self.agent_domain}")

        function = self.identity_registry.functions.newAgent(
            self.agent_domain, self.address
        )

        # Check balance
        balance = self.w3.eth.get_balance(self.address)
        required = self.w3.to_wei(0.01, "ether")
        if balance < required:
            raise ValueError(
                f"Insufficient balance. Have {self.w3.from_wei(balance, 'ether')} ETH, "
                f"need at least {self.w3.from_wei(required, 'ether')} ETH"
            )

        # Build transaction
        transaction = function.build_transaction(
            {
                "from": self.address,
                "gas": 200000,
                "gasPrice": self.w3.eth.gas_price,
                "nonce": self.w3.eth.get_transaction_count(self.address),
                "value": self.w3.to_wei(0.005, "ether"),
            }
        )

        # Sign and send
        signed_txn = self.w3.eth.account.sign_transaction(
            transaction, private_key=self.private_key
        )
        tx_hash = self.w3.eth.send_raw_transaction(signed_txn.raw_transaction)

        print(f"   Transaction hash: {tx_hash.hex()}")

        # Wait for confirmation
        receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)

        if receipt.status == 1:
            # Parse event logs
            try:
                logs = self.identity_registry.events.AgentRegistered().process_receipt(
                    receipt
                )
                if logs and len(logs) > 0:
                    agent_id = logs[0]["args"]["agentId"]
                    print(f"✅ Agent registered successfully with ID: {agent_id}")
                    self.agent_id = agent_id
                    return self.agent_id
            except Exception:
                pass

            # Fallback: query by address
            import time
            time.sleep(0.5)
            agent_info = self.identity_registry.functions.resolveByAddress(
                self.address
            ).call()
            if agent_info[0] > 0:
                self.agent_id = agent_info[0]
                print(f"✅ Agent registered successfully with ID: {self.agent_id}")
                return self.agent_id

            raise Exception("Registration succeeded but couldn't determine agent ID")
        else:
            raise Exception("Agent registration failed")

    def request_validation(self, validator_agent_id: int, data_hash: bytes) -> str:
        """
        Request validation from a validator agent

        Args:
            validator_agent_id: ID of the validator agent
            data_hash: Hash of the data to be validated

        Returns:
            Transaction hash
        """
        if not self.agent_id:
            raise ValueError("Agent must be registered first")

        print(f"🔍 Requesting validation from agent {validator_agent_id}")

        function = self.validation_registry.functions.validationRequest(
            validator_agent_id, self.agent_id, data_hash
        )

        transaction = function.build_transaction(
            {
                "from": self.address,
                "gas": 150000,
                "gasPrice": self.w3.eth.gas_price,
                "nonce": self.w3.eth.get_transaction_count(self.address),
            }
        )

        signed_txn = self.w3.eth.account.sign_transaction(
            transaction, private_key=self.private_key
        )
        tx_hash = self.w3.eth.send_raw_transaction(signed_txn.raw_transaction)

        receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)

        if receipt.status == 1:
            print(f"✅ Validation request successful")
            return tx_hash.hex()
        else:
            raise Exception("Validation request failed")

    def submit_validation_response(self, data_hash: bytes, response: int) -> str:
        """
        Submit a validation response (for validator agents)

        Args:
            data_hash: Hash of the validated data
            response: Validation score (0-100)

        Returns:
            Transaction hash
        """
        if not self.agent_id:
            raise ValueError("Agent must be registered first")

        print(f"📊 Submitting validation response: {response}/100")

        function = self.validation_registry.functions.validationResponse(
            data_hash, response
        )

        transaction = function.build_transaction(
            {
                "from": self.address,
                "gas": 120000,
                "gasPrice": self.w3.eth.gas_price,
                "nonce": self.w3.eth.get_transaction_count(self.address),
            }
        )

        signed_txn = self.w3.eth.account.sign_transaction(
            transaction, private_key=self.private_key
        )
        tx_hash = self.w3.eth.send_raw_transaction(signed_txn.raw_transaction)

        receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)

        if receipt.status == 1:
            print(f"✅ Validation response submitted successfully")
            return tx_hash.hex()
        else:
            raise Exception("Validation response submission failed")

    def get_agent_info(self, agent_id: int) -> Dict[str, Any]:
        """Get information about an agent from the registry"""
        result = self.identity_registry.functions.getAgent(agent_id).call()
        return {
            "agent_id": result[0],
            "agent_domain": result[1],
            "agent_address": result[2],
        }

File: agents/base-agent.ts
/**
 * Base Agent - Minimal ERC-8004 Implementation
 */

import {
  createPublicClient,
  createWalletClient,
  http,
  parseEther,
  formatEther,
  type Address,
  type Hash,
} from "viem";
import { privateKeyToAccount } from "viem/accounts";
import { foundry } from "viem/chains";
import { readFileSync } from "fs";
import { join } from "path";

// ============ Types ============

interface DeployedContracts {
  contracts: {
    IdentityRegistry: Address;
    ValidationRegistry: Address;
    ReputationRegistry: Address;
  };
}

// ============ Base Agent ============

export class ERC8004BaseAgent {
  protected account: ReturnType<typeof privateKeyToAccount>;
  protected publicClient: ReturnType<typeof createPublicClient>;
  protected walletClient: ReturnType<typeof createWalletClient>;

  protected identityRegistryAddress: Address;
  protected validationRegistryAddress: Address;
  protected reputationRegistryAddress: Address;

  protected identityRegistryAbi: readonly unknown[];
  protected validationRegistryAbi: readonly unknown[];
  protected reputationRegistryAbi: readonly unknown[];

  public agentId: bigint | null = null;
  public address: Address;

  constructor(public agentDomain: string, privateKey: `0x${string}`) {
    // Create account
    this.account = privateKeyToAccount(privateKey);
    this.address = this.account.address;

    // Create clients
    const rpcUrl = process.env.RPC_URL || "http://127.0.0.1:8545";

    this.publicClient = createPublicClient({
      chain: foundry,
      transport: http(rpcUrl),
    });

    this.walletClient = createWalletClient({
      account: this.account,
      chain: foundry,
      transport: http(rpcUrl),
    });

    // Load contracts
    const deployed = this.loadContracts();
    this.identityRegistryAddress = deployed.contracts.IdentityRegistry;
    this.validationRegistryAddress = deployed.contracts.ValidationRegistry;
    this.reputationRegistryAddress = deployed.contracts.ReputationRegistry;

    // Load ABIs
    this.identityRegistryAbi = this.loadAbi("IdentityRegistry");
    this.validationRegistryAbi = this.loadAbi("ValidationRegistry");
    this.reputationRegistryAbi = this.loadAbi("ReputationRegistry");
  }

  // ============ Private Methods ============

  private loadContracts(): DeployedContracts {
    const path = join(process.cwd(), "deployed_contracts.json");
    return JSON.parse(readFileSync(path, "utf-8"));
  }

  private loadAbi(contractName: string): readonly unknown[] {
    const path = join(
      process.cwd(),
      "contracts/out",
      `${contractName}.sol`,
      `${contractName}.json`
    );
    const artifact = JSON.parse(readFileSync(path, "utf-8"));
    return artifact.abi;
  }

  private async checkRegistration(): Promise<void> {
    try {
      // Check if this address owns any agent NFTs
      const totalAgents = (await this.publicClient.readContract({
        address: this.identityRegistryAddress,
        abi: this.identityRegistryAbi,
        functionName: "totalAgents",
      })) as bigint;

      // Try to find an agent owned by this address
      // Note: This is a simple check. In production, you'd want to index ownership
      for (let i = 1n; i <= totalAgents && i <= 100n; i++) {
        try {
          const owner = (await this.publicClient.readContract({
            address: this.identityRegistryAddress,
            abi: this.identityRegistryAbi,
            functionName: "ownerOf",
            args: [i],
          })) as Address;

          if (owner.toLowerCase() === this.address.toLowerCase()) {
            this.agentId = i;
            console.log(`✅ Agent already registered with ID: ${this.agentId}`);
            return;
          }
        } catch {
          // Agent ID doesn't exist or no owner
          continue;
        }
      }

      console.log("ℹ️  Agent not yet registered");
    } catch {
      console.log("ℹ️  Agent not yet registered");
    }
  }

  // ============ Public Methods ============

  async registerAgent(): Promise<bigint> {
    // First, check if already registered
    await this.checkRegistration();
    if (this.agentId !== null) return this.agentId;

    console.log(`📝 Registering agent: ${this.agentDomain}`);

    // Register with tokenURI (using agentDomain as simple identifier)
    // In production, tokenURI would point to actual metadata JSON
    const tokenURI = `ipfs://agent/${this.agentDomain}`;

    const hash: Hash = await (this.walletClient as any).writeContract({
      address: this.identityRegistryAddress,
      abi: this.identityRegistryAbi,
      functionName: "register",
      args: [tokenURI],
    });

    const receipt = await this.publicClient.waitForTransactionReceipt({ hash });

    // Parse the Registered event to get the agentId
    // Event signature: Registered(uint256 indexed agentId, string tokenURI, address indexed owner)
    // This creates a keccak256 hash of: keccak256("Registered(uint256,string,address)")
    const registeredEventSig =
      "0x4fdc8ad5e7b44eac5c54d8b9c1e7a890f8f3c9e5a2d3f3e2d1c0b0a09fef1234"; // Placeholder - will search for actual event

    // Find the Registered event (not the Transfer event from ERC-721)
    // The Transfer event signature is 0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef
    const transferEventSig =
      "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef";

    // The Transfer event for ERC-721 minting has: [eventSig, from (0x0), to, tokenId]
    // We can extract the tokenId from topics[3] of the Transfer event
    const transferEvent = receipt.logs.find((log) => {
      try {
        return (
          log.address.toLowerCase() ===
            this.identityRegistryAddress.toLowerCase() &&
          log.topics[0] === transferEventSig &&
          log.topics[1] ===
            "0x0000000000000000000000000000000000000000000000000000000000000000" // from address is 0x0 (minting)
        );
      } catch {
        return false;
      }
    });

    if (!transferEvent || !transferEvent.topics[3]) {
      console.error(
        "All logs:",
        JSON.stringify(
          receipt.logs.map((l) => ({
            address: l.address,
            topics: l.topics,
          })),
          null,
          2
        )
      );
      throw new Error(
        "Failed to find Transfer (minting) event in transaction receipt"
      );
    }

    // The tokenId (agentId) is in topics[3] of the Transfer event
    this.agentId = BigInt(transferEvent.topics[3]);

    console.log(`✅ Registered with ID: ${this.agentId}`);
    return this.agentId;
  }

  async requestValidation(
    validatorAddress: Address,
    requestUri: string,
    requestHash: `0x${string}`
  ): Promise<Hash> {
    const hash = await (this.walletClient as any).writeContract({
      address: this.validationRegistryAddress,
      abi: this.validationRegistryAbi,
      functionName: "validationRequest",
      args: [validatorAddress, this.agentId, requestUri, requestHash],
    });

    await this.publicClient.waitForTransactionReceipt({ hash });
    console.log("✅ Validation request successful");
    return hash;
  }

  async submitValidationResponse(
    requestHash: `0x${string}`,
    response: number,
    responseUri: string = "",
    responseHash: `0x${string}` = "0x0000000000000000000000000000000000000000000000000000000000000000",
    tag: `0x${string}` = "0x0000000000000000000000000000000000000000000000000000000000000000"
  ): Promise<Hash> {
    const hash = await (this.walletClient as any).writeContract({
      address: this.validationRegistryAddress,
      abi: this.validationRegistryAbi,
      functionName: "validationResponse",
      args: [requestHash, response, responseUri, responseHash, tag],
    });

    await this.publicClient.waitForTransactionReceipt({ hash });
    console.log("✅ Validation response submitted");
    return hash;
  }

  async getBalance(): Promise<string> {
    const balance = await this.publicClient.getBalance({
      address: this.address,
    });
    return formatEther(balance);
  }
}
